actix-files/src/path_buf.rs

```diff
@@ -26,6 +26,10 @@ impl PathBufWrap {
     pub fn parse_path(path: &str, hidden_files: bool) -> Result<Self, UriSegmentError> {
         let mut buf = PathBuf::new();
 
+        let path = percent_encoding::percent_decode_str(path)
```

Comment:
Paths in URLs are lists of percent-encoded components, not percent-encoded lists of components. Decoding an entire path string can introduce new components by decoding forward/back-slashes.

For example: `https://example.com/music/bands/AC%2FDC`

Decode then split results in the components: `["music", "bands", "AC", "DC"]`
Split then decode results in the correct components: `["music", "bands", "AC/DC"]`

If you have a component like `foo%2F..%2F..%2F..`, decode-then-split would introduce new ".." components, which can be a security risk as it defeats any previous validation of the path.

Reply:
Indeed I'm aware of this https://github.com/actix/actix-web/pull/2398#discussion_r726706854.

What is arguable though is the assumption that there is a one-to-one mapping between url path components and filesystem path components. In my opinion, a file server can safely interpret a single url path component to be multiple filesystem  path components as long as this transformation is done before any validation step (which is the case here). So I don't really see a problem here.

Reply:
Ah - I missed your comment about it. There's no standard describing how this conversion should be done, and all kinds of funky edge-cases, so validating the result is definitely a good idea 👍.

FWIW, Chromium [will reject](https://github.com/chromium/chromium/blob/87ca5380f1d05f7218edd5659ed857b16a0bb2d0/net/base/filename_util.cc#L59) such URLs from being turned in to file paths. In my own library (for the Swift language), I also reject percent-encoded NULL bytes, because ultimately you're likely to be talking to the OS using null-terminated strings. Windows paths are significantly more complex with all kinds of additional edge-cases.

Reply:
I followed [the discussion](https://crbug.com/585422) for chrome and couldn't see any security argument being made to justify the behavior. Meanwhile, firefox, nginx and `python -m http.server` do decode `%2F` into `/` for file paths. So I guess we're on the safe side!

For NULL chars, we rely on the std lib to reject it.

Reply:
@karwa thanks for your insight on this 👍🏻